### PR TITLE
Use native line numbers in Emacs 26

### PR DESCRIPTION
**Pull request by bmag**
Note to other maintainers: don't merge this yet, I might need to make a few more changes

Emacs 26.1 will introduce a new minor mode for native line numbers: `display-line-numbers-mode`. It is meant to replace linum and friends (including linum-relative). This PR causes Spacemacs to use `display-line-numbers-mode` in Emacs 26, and `linum-mode` in Emacs 25. If `nlinum` layer is enabled by the user, then `nlinum-mode` will be used instead of `display-line-numbers-mode`. Users of `nlinum` layer are recommended to use `display-line-numbers-mode` in Emacs 26.

There's a small issue that `display-line-numbers-type` is a "global" variable (not buffer-local), so toggling between absolute and relative line numbers in one buffer may affect other buffers. If you test this PR and want to comment about how it behaves, please do.

Currently, the `SPC t n` and `SPC t r` toggles for the three different line numbers solutions behave differently. I'm not sure which behavior is the best, so I'd like to hear what other people prefer. The different behaviors are as follows:

(this summary shows state transitions for line numbers status. "off" means no line numbers are shown, "absolute" means absolute line numbers are shown, "relative" means relative line numbers are shown)
linum:
- `SPC t n`: (state before key press --> state after key press)
  - off (variant 1) --> absolute
  - off (variant 2) --> relative
  - absolute --> off (variant 1)
  - relative --> off (variant 2)
- `SPC t r`:
  - off (variant 1) --> relative
  - off (variant 2) --> off (variant 1)
  - absolute --> relative
  - relative --> off (variant 1)

nlinum:
- `SPC t n`:
  - off (variant 1) --> absolute
  - off (variant 2) --> relative
  - absolute --> off (variant 1)
  - relative --> off (variant 2)
- `SPC t r`:
  - off (variant 1) --> off (variant 2)
  - off (variant 2) --> off (variant 1)
  - absolute --> relative
  - relative --> absolute

display-line-numbers:
- `SPC t n`:
  - off --> absolute
  - absolute --> off
  - relative --> absolute
- SPC t r:
  - off --> relative
  - absolute --> relative
  - relative --> off

### CODE REVIEW

**Comment by jsoo1** [comment]
Is there any way to keep the same state transitions in native line numbers as in nlinum?

**Comment by jsoo1** [comment]
Also, awesome work and thank you.

**Comment by jsoo1** [comment]
So far it looks good.  I was able to use relative numbers in a large file with no problem.  Tested on GUI and daemon (frame and terminal).  Both look good. Some feedback:
  1. The changes to the state transitions are a little jarring. I would prefer if we could keep the old transitions.
  1.  Some modes that don't normally have line numbers have line numbers (Magit, in particular).

**Comment by syl20bnr** [comment]
bmag <notifications@github.com> writes:

```
display-line-numbers:
- `SPC t n`:
  - off --> absolute
  - absolute --> off
  - relative --> absolute
- SPC t r:
  - off --> relative
  - absolute --> relative
  - relative --> off
```

This seems the most natural and as it should be since beginning.
I'm not found of the current toggle behavior in develop, does not feel POLA.

> Users of `nlinum` layer are recommended to use `display-line-numbers-mode` in Emacs 26.

Let's deprecate it for Emacs 26+ users, there is no point in using it
right ?

**Comment by syl20bnr** [comment]
And let's make the relative toggle key binding be `SPC t N` instead of `SPC t r` :-)

**Comment by jsoo1** [comment]
Aww man! I liked `SPC t r`...

**Comment by bmag** [comment]
> Some modes that don't normally have line numbers have line numbers (Magit, in particular).

Because we have a bug: setting `dotspacemacs-line-numbers` to `t` or `relative` enables them in all modes, instead of only in `prog-mode` and `text-mode` derivatives (as the documentation claims). I have a fix for it that I will push later.

By the way, turns out `magit-mode` itself disables `linum-mode` and `nlinum-mode` in Magit buffers. If it didn't do that, the same bug would happen with linum-mode as well.

>> Users of `nlinum` layer are recommended to use `display-line-numbers-mode` in Emacs 26.

> Let's deprecate it for Emacs 26+ users

You sure about it? I didn't deprecate the layer because it is opt-in, but I'm ok with deprecating it if you're sure.

File: layers/+spacemacs/spacemacs-defaults/packages.el [hunk 1]
```diff
@@ -17,14 +17,16 @@
         (conf-mode :location built-in)
         (dired :location built-in)
         (dired-x :location built-in)
+        (display-line-numbers :location built-in
+                              :toggle (version<= "26" emacs-version))
```
[the remark this answers is not in the document]
Comment (bmag) [reply layers/+spacemacs/spacemacs-defaults/packages.el]: Doesn't really matter. The first 26.x release will be 26.1, and the development process already reached RC1 where`emacs-version` says "26.1". Any Emacs build that calls itself version 26.0.x is old and irrelevant by now.

**Comment by bmag** [comment]
Made a small correction to the toggles (so toggling in one buffer won't affect state of other buffers) and fixed the bug that @jsoo1 and I found (see https://github.com/syl20bnr/spacemacs/pull/10675/commits/34ba6052725070e9dfe60a7823fbda20c2421a17 for details).

cc @deb0ch

**Comment by deb0ch** [comment]
@bmag you called me ?

So excited about this PR, learning at the same time the existence of native line numers in Emacs 26 and that Spacemacs will integrate them 😺

File: layers/+spacemacs/spacemacs-defaults/packages.el (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsoo1) [reply layers/+spacemacs/spacemacs-defaults/packages.el]: Ok. I just want to make sure hat if someone decided to use some version 26.0.x in the future for any reason it wouldn’t break for them.

**Comment by syl20bnr** [comment]
If it does not bring any benefit for Emacs 26 users, there is no need to
make it available post Emacs 26.

Should be still usable for users with Emacs 25 though.

**Comment by kommen** [comment]
I'm testing this in my spacemacs config with Emacs 26.1rc1 and so far it works great, thank you @bmag!
As a user, the state transition mappings work for me.
I just noticed one very minor issue though: when starting emacs, on the `*spacemacs*` buffer the line numbers a briefly visible, leading to a line wrap in the header. See

screen shot 2018-05-21 at 18 35 59

**Comment by bmag** [comment]
> when starting emacs, on the \*spacemacs\* buffer the line numbers a briefly visible, leading to a line wrap in the header.

I also see them enabled briefly during startup, and after a short while they are disabled (before startup finishes). Don't know why yet. Not sure what you meant about line wrap, though, isn't it normal (except the line numbers)?

**Comment by kommen** [comment]
@bmag yes, after a short while they are disabled again for me as well.

I mean that while the line numbers are visible, the "emacs)" part of "(spacemacs)" in the top right wraps, as visible in the screenshot.  This seems to be a result of the right-aligning not respecting the width of the line numbers?

**Comment by bmag** [comment]
@kommen ok, now I see it. The native line numbers use their own new margin-like space, maybe the home buffer's code doesn't respect that.

Anyway, I pushed a fix for it now, thanks.

**Comment by kommen** [comment]
@bmag thanks, also fixes it for me.

**Comment by rekols** [comment]
wow, this is fast !!!

**Comment by rekols** [comment]
wow, this is fast !!!

**Comment by syl20bnr** [comment]
There is no support for this in the layer system, I just put a dirty
conditional block in
https://github.com/syl20bnr/spacemacs/commit/8660f93de :-)

**Pull request closed without merging**